Support setting a custom window size in Writer

arl opened the pull request:
In #7 this adds support to custom window size (called WindowLog in the code, as in libzstd).

As discussed, setting this parameter is performed via a `WriterParams` structure in which letting all values to their default is equivalent to calling `NewWriter`.

I took the liberty to modify `newWriterParams` the non-public Writer constructor so that it now takes a `WriterParam` instead of having to add a new parameter to it.
One example and 1 test have been added.

Fixes #7

arl commented:
I previously added a test `TestWriterWindowLogGreaterThan27` but travis worker kills it 2 times over 3 because of increased memory.

valyala commented on writer.go:
```diff
@@ -65,12 +70,60 @@ func NewWriterLevel(w io.Writer, compressionLevel int) *Writer {
+	// decompressor requires special treatment.
+	WindowLog int
+
+	// Dict is the dictionnary used for compression. May be nil.
```
Could you update this line to `Dict is optional dictionary used for compression`?

arl replied:
done

valyala commented on writer.go:
```diff
@@ -65,12 +70,60 @@ func NewWriterLevel(w io.Writer, compressionLevel int) *Writer {
+	return newWriterParams(w, &params)
+}
+
+func newWriterParams(w io.Writer, params *WriterParams) *Writer {
```
The `newWriterParams` function body may be inlined into `NewWriterParams` func.

arl replied:
done

valyala commented on writer.go:
```diff
@@ -65,12 +70,60 @@ func NewWriterLevel(w io.Writer, compressionLevel int) *Writer {
+// to finalize the compressed stream.
+//
+// Call Release when the Writer is no longer needed.
+func NewWriterParams(w io.Writer, params WriterParams) *Writer {
```
I'd suggest passing `params` by reference, so it could accept `nil` for default params.

arl replied:
done

valyala commented on writer.go:
```diff
@@ -99,27 +153,35 @@ func newWriterDictLevel(w io.Writer, cd *CDict, compressionLevel int) *Writer {
 // Reset resets zw to write to w using the given dictionary cd and the given
-// compressionLevel.
+// compressionLevel. Other parameters set via WriterParams remain unchanged.
 func (zw *Writer) Reset(w io.Writer, cd *CDict, compressionLevel int) {
```
I'd suggest adding `ResetWriterParams(w io.Writer, params *WriterParams)` function as well for updating arbitrary set of params.

arl replied:
done

valyala commented on writer.go:
```diff
@@ -99,27 +153,35 @@ func newWriterDictLevel(w io.Writer, cd *CDict, compressionLevel int) *Writer {
 
 	zw.cd = cd
-	initCStream(zw.cs, zw.cd, compressionLevel)
+	params := &WriterParams{
```
Could you verify whether this line leads to memory allocation? If yes, then I'd suggest passing `params` by value to `initCStream` in order to remove this memory allocation, since `Writer.Reset` is usually used in functions sensitive for performance, which want minimizing memory allocations.

arl replied:
in `Reset` actually it doesn't.
But as you suggested I added `ResetWriterParams` method and this line does allocate even when passing params to `initCStream`, but I'm not sure there's an easy way around that since, as you can see, `params` escapes to heap because of the of the assignment of `params.Dict` `to the receveir's `cd`, namely  `zw.cd`.

```
./writer.go:174:50: parameter params leaks to {heap} with derefs=1:
./writer.go:174:50:   flow: {heap} = *params:
./writer.go:174:50:     from params.Dict (dot of pointer) at ./writer.go:180:16
./writer.go:174:50:     from zw.cd = params.Dict (assign) at ./writer.go:180:8
```
Passing `params` by value or pointer to `initCStream` doesn't change that fact.

Let me know if you see something else we could try to avoid that allocation

valyala commented on writer.go:
```diff
@@ -99,27 +153,35 @@ func newWriterDictLevel(w io.Writer, cd *CDict, compressionLevel int) *Writer {
 		ensureNoError("ZSTD_initCStream", result)
 	}
+
+	result := C.ZSTD_CCtx_setParameter(cs, C.ZSTD_cParameter(C.ZSTD_c_windowLog), C.int(params.WindowLog))
```
Do we need adjusting `params.WindowLog` to default value when it is set to 0 or `zstd` properly handles zero value for this arg?

arl replied:
`zstd` properly handles the zero value.

https://github.com/facebook/zstd/blob/5b0a452cac9672a3003781541235dc47727e6c2a/lib/zstd.h#L284-L291

arl commented on writer.go:
```diff
@@ -54,7 +55,11 @@ func NewWriter(w io.Writer) *Writer {
+	params := &WriterParams{
+		CompressionLevel: compressionLevel,
+	}
+
```
done

arl commented on writer.go:
```diff
@@ -43,7 +44,7 @@ type Writer struct {
 // Call Release when the Writer is no longer needed.
 func NewWriter(w io.Writer) *Writer {
-	return newWriterDictLevel(w, nil, DefaultCompressionLevel)
+	return newWriterParams(w, &WriterParams{})
```
done

arl commented on writer.go:
```diff
@@ -65,12 +70,60 @@ func NewWriterLevel(w io.Writer, compressionLevel int) *Writer {
+// Calling NewWriterParams with a zero-value WriterParams is equivalent to
+// calling NewWriter.
+type WriterParams struct {
+	// Compression level. Special value 0 means default.
```
done

arl commented on writer.go:
```diff
@@ -65,12 +70,60 @@ func NewWriterLevel(w io.Writer, compressionLevel int) *Writer {
+// A WriterParams allows users to specify compression parameters by calling
+// NewWriterParams.
+//
+// Calling NewWriterParams with a zero-value WriterParams is equivalent to
```
done

arl commented:
@valyala I believe I addressed your comments. There's an allocation in `ResetWriterParams` but not in `ResetWriter`.
I see the coverage fails, I can add a test for `ResetWriterParams`

valyala commented on writer.go:
```diff
@@ -99,27 +151,49 @@ func newWriterDictLevel(w io.Writer, cd *CDict, compressionLevel int) *Writer {
+// compressionLevel. Use ResetWriterParams if you wish to change other
+// parameters that were set via WriterParams.
 func (zw *Writer) Reset(w io.Writer, cd *CDict, compressionLevel int) {
 	zw.inBuf.size = 0
```
Could you init `params` and then call `ResetWriterParams` here? This will remove duplicate code in the `Reset` and `ResetWithParams` functions.

arl replied:
done

valyala commented:
> @valyala I believe I addressed your comments. There's an allocation in `ResetWriterParams` but not in `ResetWriter`.

It is unclear why `ResetWriterParams` allocates if `params` is passed by value into `initCStream`. Could you add a benchmark, which calls `Reset` and `ResetWriterParams` in a loop and see whether it really allocates by calling [ReportAllocs()](https://golang.org/pkg/testing/#B.ReportAllocs) before the loop?

> I see the coverage fails, I can add a test for `ResetWriterParams`

This would be nice to have.

arl commented:
Indeed neither `Reset` nor `ResetWriterParams` allocate. I let the benchmark in `writer_timing_test.go`

valyala commented:
@arl , thanks for the contribution!
